### PR TITLE
feat: email notifications for supporter badge

### LINKED ISSUE #2688: [feature] badges add transactional email notifications for users and admins
Precious plastic to provide listing of required transactional emails, their trigger and the email content.

### PR DESCRIPTION
**Pull request by thisislawatts**
PR Checklist

- [x] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Developer experience (improves developer workflows for contributing to the project)

## Description

Adds email notification when supporter badge is added or removed.

- feat: adds email notification when supporter badge added/removed
- feat(emails): load HTML template
- chore: add local server for rendering
- feat: move supporter-badge-added to hbs template
- chore(email): add title to preview app
- fix: attempt to override some type warnings

## Git Issues

Partial fix #2688

---

## What happens next?

Thanks for the contribution! We try to make sure all PRs are reviewed ahead of a monthly dev call (first Monday of the month, open to all!).

If the PR is working as intended it'll be merged and included in the next platform release, if not changes will be requested and re-reviewed once updated.

If you need more immediate feedback you can try reaching out on Discord in the [Community Platform `development` channel](https://discord.com/channels/586676777334865928/938781727017558018).

### CODE REVIEW

File: functions/src/emailNotifications/README.md [hunk 1]
```diff
@@ -62,7 +62,8 @@ We do not have the email extension set up in the emulator environment, so testin
 
 1. Make changes to the function.
 2. Use the [unit tests](./createNotificationEmails.spec.ts) to test new transformation functionality and generate new html snapshots.
-3. To preview the new email, either copy/paste the html snapshot into a [free online site](https://codebeautify.org/htmlviewer) or into new file and use a VSCode extension like HTML Preview. Include a screenshot in your PR description.
+
+Additionally you can run `yarn serve:email-templates` to run a local vite server that will serve the email templates. This is useful for testing the email templates in a browser.
```
Comment (evakill) [inline functions/src/emailNotifications/README.md:66]:
```suggestion
Additionally you can run `yarn serve:email-templates` to run a local vite server that will serve the email templates. This is useful for testing the email templates in a browser. You can find the code for this in the [development](./development) directory.
```

File: functions/src/emailNotifications/supporterBadgeEmails.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        if (
+          before.badges?.supporter === true &&
+          after.badges?.supporter === false
+        ) {
+          await db.collection(DB_ENDPOINTS.emails).add({
+            to: toUserEmail,
+            message: getUserSupporterBadgeRemovedEmail(after as IUserDB),
+          })
+        }
+      },
```
Comment (evakill) [inline functions/src/emailNotifications/supporterBadgeEmails.ts:41]: nits: 
1. preference to split this into its own fn, i.e. `handleUserSupporterBadgeChange`, for readability
2. you can also add typing to the `change` param, `Change<QueryDocumentSnapshot<IUserDB>>`, which would avoid the casting on 28 and 38
3. the param on 14 and then passed in on 24, `snapshot`, is better named `change`, as they are different firebase concepts in my understanding

File: functions/src/emailNotifications/development/entry-server.mjs
Comment (evakill) [inline functions/src/emailNotifications/development/entry-server.mjs:1]: this is so cool and helpful!!!! thank you

File: functions/src/emailNotifications/index.ts [hunk 3]
```diff
@@ -119,3 +120,8 @@ exports.sendMapPinSubmissionEmail = functions.firestore
   .onCreate((snapshot, context) =>
     withErrorAlerting(context, createMapPinSubmissionEmail, [snapshot.data()]),
   )
+
+/** Watch new user docs and trigger emails on supporter */
+exports.sendSupporterEmail = (
+  supporterBadgeEmails as any
+).handleUserSupporterBadgeUpdate
```
Comment (evakill) [inline functions/src/emailNotifications/index.ts:127]: nice, i think good idea to use the same file org / import pattern for the other email groupings as this service grows 👍

File: functions/src/emailNotifications/supporterBadgeEmails.spec.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  },
+}))
+
+describe('supporterBadgeEmails', () => {
+  beforeEach(async () => {
+    await FirebaseEmulatedTest.clearFirestoreDB()
+  })
+
+  it('should not send an email if supporter badge status unchanged', async () => {
+    const db = FirebaseEmulatedTest.admin.firestore()
```
Comment (evakill) [inline functions/src/emailNotifications/supporterBadgeEmails.spec.ts:20]: ok to put this at the 'describe' level i think

File: functions/src/emailNotifications/supporterBadgeEmails.spec.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    getUser: () => ({
+      email: 'test@test.com',
+    }),
+  },
+}))
+
+describe('supporterBadgeEmails', () => {
+  beforeEach(async () => {
+    await FirebaseEmulatedTest.clearFirestoreDB()
+  })
```
Comment (evakill) [inline functions/src/emailNotifications/supporterBadgeEmails.spec.ts:17]: seems like we have a pattern of clearing the DB before and after the tests. i'm not super sure this is necessary, but if we wanted to be consistent we could add a clear in an `afterAll` hook

File: functions/src/emailNotifications/supporterBadgeEmails.spec.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+  it('should not send an email if supporter badge status unchanged', async () => {
+    const db = FirebaseEmulatedTest.admin.firestore()
+    const wrapped = fbTest.wrap(fun.handleUserSupporterBadgeUpdate)
```
Comment (evakill) [inline functions/src/emailNotifications/supporterBadgeEmails.spec.ts:21]: ah! this was the function i was looking for. i couldn't get the test emulator to respond to DB changes without it... and now i know why. good find. i will update the other tests to use this as well.

File: functions/src/emailNotifications/supporterBadgeEmails.spec.ts [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            badges: {
+              supporter: true,
+            },
+          },
+          'v3_users/123',
+        ),
+      ),
+    )
+
+    // Assert
```
Comment (evakill) [inline functions/src/emailNotifications/supporterBadgeEmails.spec.ts:47]: are these "Act" and "Assert" unfinished comments?

**Review by evakill — COMMENTED** [review]
woahh this is so great, thanks luke! i love the idea of the preview server and you've introduced some super solid patterns that i'd like to integrate into the rest of the email work. i left some really minor comments but i'm excited to ship this

File: functions/src/emailNotifications/supporterBadgeEmails.spec.ts (hunk 7 above)
Comment (thisislawatts) [reply functions/src/emailNotifications/supporterBadgeEmails.spec.ts:47]: Ah no, these are some formatting comments I use to visually separate the test code:
https://automationpanda.com/2020/07/07/arrange-act-assert-a-pattern-for-writing-good-tests/

File: functions/src/emailNotifications/supporterBadgeEmails.ts (hunk 2 above)
Comment (thisislawatts) [reply functions/src/emailNotifications/supporterBadgeEmails.ts]: Fixed in: https://github.com/ONEARMY/community-platform/pull/2812/commits/329e416b3c0c6060af6a3254a8f2bbf8d491ec47

File: functions/src/emailNotifications/supporterBadgeEmails.spec.ts (hunk 5 above)
Comment (thisislawatts) [reply functions/src/emailNotifications/supporterBadgeEmails.spec.ts]: Good catch thanks, added in: https://github.com/ONEARMY/community-platform/pull/2812/commits/a1f19ab61af0dfd2c6657f2574bac6926d03c7d4

File: functions/src/emailNotifications/supporterBadgeEmails.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  change: functions.Change<QueryDocumentSnapshot<IUserDB>>,
+) => {
+  const after = change.after.exists ? change.after.data() : null
+  const before = change.before.exists ? change.before.data() : null
+  const toUserEmail = after ? await getUserEmail(after._authID) : null
+
+  if (before.badges?.supporter === false && after.badges?.supporter === true) {
+    await db.collection(DB_ENDPOINTS.emails).add({
+      to: toUserEmail,
+      message: getUserSupporterBadgeAddedEmail(after as IUserDB),
```
Comment (evakill) [inline functions/src/emailNotifications/supporterBadgeEmails.ts:29]: can remove cast here and on 36

**Review by evakill — APPROVED** [review]
thanks for changes!

**Pull request merged**